### ISSUE TITLE
Add centralized logging

### ISSUE THREAD

**Issue by Thomsen96** [DevelOpsITU/MiniTwit#126]
From the discussion/ design decision in #125. We want to add centralized logging

**Comment by Thomsen96** [comment]
I have experimented a little bit with elastic search. And it seams like it will take alot of resources. 

When starting it locally it took 5GB of ram, and then i ran out. and it also complained that it needed more with the error message:
```shell
elasticsearch  | ERROR: [1] bootstrap checks failed. You must address the points described in the following [1] lines before starting Elasticsearch.
elasticsearch  | bootstrap check failure [1] of [1]: max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]
```

I am shifting towards  Loki and continue experimenting.

**Comment by Thomsen96** [comment]
TLDR in bottom. 

After ~6 hours and commit https://github.com/DevelOpsITU/MiniTwit/commit/e8b255870b0de1312488597e0afffe6714071360

did I get the setup to work with the docker logging driver fluentd to fluentbit to loki to work. 

![billede](https://user-images.githubusercontent.com/32035863/159163409-e5045c9c-db9d-4d96-8c04-83f75a07e7a0.png)


The big headache was that when using the docker logging driver fulentd, it sent the log in the format: 

```json
{
    "container_id":"c7ba0295c842383cc3499adae6646a6a871af01c26c37ce349cc75ef87ec920b"
    "container_name":"/minitwit"
    "log":{"level":"error","error":"failed to connect to `host=postgres user=postgres database=minitwit`: server error (FATAL: the database system is starting up (SQLSTATE 57P03))","time":"2022-03-20T12:24:24Z","message":"Could not open database"}
    "source":"stderr"
```

Where it is only the key `log` that actually comes from the application. and when it is sent to loki, is the entire json object not sent like a json object but as a string with escaped qoutes (\\") instead of (\")
```json
{
    "container_id":"c7ba0295c842383cc3499adae6646a6a871af01c26c37ce349cc75ef87ec920b"
    "container_name":"/minitwit"
    "log":"{\"level\":\"error\",\"error\":\"failed to connect to `host=postgres user=postgres database=minitwit`: server error (FATAL: the database system is starting up (SQLSTATE 57P03))\",\"time\":\"2022-03-20T12:24:24Z\",\"message\":\"Could not open database\"}"
    "source":"stderr"
```

also seen/debugged it with wireshark:
![billede](https://user-images.githubusercontent.com/32035863/159164596-61873cef-260d-4848-a9cb-451fbc777099.png)




In the end here I got it working so that it parses the data in the json key `log` and can use the other keys as log labels in Loki.

TLDR:

- bad: 
![billede](https://user-images.githubusercontent.com/32035863/159164487-51263774-b025-4521-9ed7-160a2eb106b2.png)
- good:
![billede](https://user-images.githubusercontent.com/32035863/159164494-dd5b3dcc-82e7-443b-903f-430a083e828a.png)

**Comment by Thomsen96** [comment]
Also regarding CPU usage. Then with the command `docker stats` is it visible that none of the services takes alot of resources

```shell
CONTAINER ID   NAME         CPU %     MEM USAGE / LIMIT     MEM %     NET I/O           BLOCK I/O   PIDS
1a016b0cfbda   minitwit     0.00%     36.34MiB / 10.69GiB   0.33%     124kB / 1.1MB     0B / 0B     14  
52d6720e37b3   fluentbit    0.02%     21.29MiB / 10.69GiB   0.19%     139kB / 180kB     0B / 0B     4   
5e741c68c644   loki         0.11%     59.97MiB / 10.69GiB   0.55%     735kB / 2.78MB    0B / 0B     21  
ad36a71cac67   postgres     0.00%     33.67MiB / 10.69GiB   0.31%     24.9kB / 27.1kB   0B / 0B     9   
4e0f37e473be   prometheus   0.00%     77.2MiB  / 10.69GiB   0.71%     22.8MB / 1.73MB   0B / 0B     22  
214701c57afc   grafana      0.08%     54.23MiB / 10.69GiB   0.50%     2.79MB / 2.72MB   0B / 0B     22  
```

**Comment by Thomsen96** [comment]
This general issue is closed. And new tasks have been created 
#157 #158
